File: building_blocks/python/knowledge_base/kb_python/kb_memory/construct_mem_db.py
from basic_contruct_db import BasicConstructDB


class ConstructMemDB(BasicConstructDB):
    def __init__(self,host,port,dbname,user,password,database):
        BasicConstructDB.__init__(self,host,port,dbname,user,password,database)
        self.kb_name = None
        self.working_kb = None
        self.composite_path = {}
        self.composite_path_values = {}
        
    

        
  
    
    def add_kb(self, kb_name, description=""):
        if kb_name in self.composite_path:
            raise ValueError(f"Knowledge base {kb_name} already exists")
        self.composite_path[kb_name] = [kb_name]
        self.composite_path_values[kb_name] = {}
        BasicConstructDB.add_kb(self, kb_name, description)
      
    def select_kb(self, kb_name):
        if kb_name not in self.composite_path:
            raise ValueError(f"Knowledge base {kb_name} does not exist")
        self.working_kb = kb_name
        
   
    def add_header_node(self, link, node_name, node_data,description=""):
        """
        Adds a header node to the knowledge base.

        Args:
            link: The link associated with the header node.
            node_name: The name of the header node.
            node_data: Data associated with the header node. Must be a dictionary.  
            description: Description of the header node.
        """
        
        
        if not isinstance(description, str):
            raise TypeError("description must be a string")
        if not isinstance(node_data, dict):
            raise TypeError("node_data must be a dictionary")
        
        if description != None:
            node_data["description"] = description
            


        self.composite_path[self.working_kb].append(link)
        self.composite_path[self.working_kb].append(node_name)
        node_path = ".".join(self.composite_path[self.working_kb])
        

        if node_path in self.composite_path_values[self.working_kb]:
            raise ValueError(f"Path {node_path} already exists in knowledge base")
        
        self.composite_path_values[self.working_kb][node_path] = True
       
        path = ".".join(self.composite_path[self.working_kb])
        print("path", path)
        BasicConstructDB.store(self, path, node_data)
       

    def add_info_node(self, link, node_name,  node_data,description=""):
        self.add_header_node(link, node_name,  node_data,description)
     
        self.composite_path[self.working_kb].pop()  # Remove node_name
        self.composite_path[self.working_kb].pop()  # Remove link
        
    
    def leave_header_node(self, label, name ):
        """
        Leaves a header node, verifying the label and name.
        If an error occurs, the knowledge_base table is deleted if it exists
        and the PostgreSQL connection is terminated.

        Args:
            label: The expected link of the header node.
            name: The expected name of the header node.
        """
        # Try to pop the expected values
        if not self.composite_path[self.working_kb]:
            raise ValueError("Cannot leave a header node: path is empty")
        
        ref_name = self.composite_path[self.working_kb].pop()
        if not self.composite_path[self.working_kb]:
            # Put the name back and raise an error
            self.composite_path[self.working_kb].append(ref_name)
            raise ValueError("Cannot leave a header node: not enough elements in path")
            
        ref_label = self.composite_path[self.working_kb].pop()
        
        # Verify the popped values
        if ref_name != name or ref_label != label:
            # Create a descriptive error message
            error_msg = []
            if ref_name != name:
                error_msg.append(f"Expected name '{name}', but got '{ref_name}'")
            if ref_label != label:
                error_msg.append(f"Expected label '{label}', but got '{ref_label}'")
                
            raise AssertionError(", ".join(error_msg))
        
        
   
    
    def check_installation(self):
        """
        Checks if the installation is correct by verifying that the path is empty.
        If the path is not empty, the knowledge_base table is deleted if present,
        the database connection is closed, and an exception is raised.
        If the path is empty, the database connection is closed normally.

        Returns:
            bool: True if installation check passed

        Raises:
            RuntimeError: If the path is not empty
        """
        for kb_name in self.composite_path:
            if len(self.composite_path[kb_name]) != 1:
                
                raise RuntimeError(f"Installation check failed: Path is not empty for knowledge base {kb_name}. Path: {self.composite_path[kb_name]}")
            if self.composite_path[kb_name][0] != kb_name:
                raise RuntimeError(f"Installation check failed: Path is not empty for knowledge base {kb_name}. Path: {self.path[kb_name]}")
                
                
                
                
                
                
if __name__ == '__main__':
    password = input("Enter PostgreSQL password: ")
    # Example Usage
    # Replace with your actual database credentials
    DB_HOST = "localhost"
    DB_PORT = "5432"
    DB_NAME = "knowledge_base"
    DB_USER = "gedgar"
    DB_PASSWORD = password
    DB_TABLE = 'composite_memory_kb'

    print("starting unit test")
    kb = ConstructMemDB(DB_HOST,DB_PORT,DB_NAME,DB_USER,DB_PASSWORD,DB_TABLE)



 
    kb.add_kb("kb1", "First knowledge base")
    kb.select_kb("kb1")
    kb.add_header_node("header1_link", "header1_name",  {"data":"header1_data"},"header1_description")
   
    kb.add_info_node( "info1_link", "info1_name",  {"data":"info1_data"},"info1_description")

    kb.leave_header_node("header1_link", "header1_name")
 
    kb.add_header_node("header2_link", "header2_name",  {"data":"header2_data"},"header2_description")
    kb.add_info_node("info2_link", "info2_name",  {"data":"info2_data"},"info2_description")
    
    kb.leave_header_node("header2_link", "header2_name")
  
    kb.add_kb("kb2", "Second knowledge base")
    kb.select_kb("kb2")
    kb.add_header_node("header1_link", "header1_name",  {"data":"header1_data"},"header1_description")
   
    
    kb.add_info_node( "info1_link", "info1_name",  {"data":"info1_data"},"info1_description")
  

    kb.leave_header_node("header1_link", "header1_name")
   
    kb.add_header_node("header2_link", "header2_name",  {"data":"header2_data"},"header2_description")
    kb.add_info_node("info2_link", "info2_name",  {"data":"info2_data"},"info2_description")
    
    kb.leave_header_node("header2_link", "header2_name")
   


    # Example of check_installation
    try:
        kb.check_installation()
        print(kb.export_to_postgres(create_table=True,clear_existing=True))
        print(kb.import_from_postgres(table_name='composite_memory_kb'))
    except RuntimeError as e:
        print(f"Error during installation check: {e}")
    print("ending unit test")

File: system_control/libraries/file_loader/test_files/kb_go/kb_memory/kb_memory_module/ConstructMemDB.go
package kb_memory_module

import (
	"fmt"
	//"log"
	"strings"
)

// ConstructMemDB extends BasicConstructDB with knowledge base management and composite path tracking
type ConstructMemDB struct {
	*BasicConstructDB                    // Embedded struct for inheritance-like behavior
	kbName              *string          // Currently selected knowledge base name
	workingKB           *string          // Working knowledge base
	compositePath       map[string][]string          // Tracks composite paths for each KB
	compositePathValues map[string]map[string]bool   // Tracks existing paths in each KB
}

// NewConstructMemDB creates a new ConstructMemDB instance
func NewConstructMemDB(host string, port int, dbname, user, password, database string) *ConstructMemDB {
	return &ConstructMemDB{
		BasicConstructDB:    NewBasicConstructDB(host, port, dbname, user, password, database),
		kbName:              nil,
		workingKB:           nil,
		compositePath:       make(map[string][]string),
		compositePathValues: make(map[string]map[string]bool),
	}
}

// AddKB adds a knowledge base with composite path tracking
func (cmdb *ConstructMemDB) AddKB(kbName, description string) error {
	// Check if KB already exists in composite path
	if _, exists := cmdb.compositePath[kbName]; exists {
		return fmt.Errorf("knowledge base %s already exists", kbName)
	}

	// Initialize composite path structures
	cmdb.compositePath[kbName] = []string{kbName}
	cmdb.compositePathValues[kbName] = make(map[string]bool)

	// Call parent method
	return cmdb.BasicConstructDB.AddKB(kbName, description)
}

// SelectKB selects a knowledge base to work with
func (cmdb *ConstructMemDB) SelectKB(kbName string) error {
	if _, exists := cmdb.compositePath[kbName]; !exists {
		return fmt.Errorf("knowledge base %s does not exist", kbName)
	}
	cmdb.workingKB = &kbName
	return nil
}

// AddHeaderNode adds a header node to the knowledge base
func (cmdb *ConstructMemDB) AddHeaderNode(link, nodeName string, nodeData map[string]interface{}, description string) error {
	if cmdb.workingKB == nil {
		return fmt.Errorf("no working knowledge base selected")
	}

	// Validate input types
	if nodeData == nil {
		return fmt.Errorf("nodeData must be a dictionary")
	}

	// Add description if provided
	if description != "" {
		nodeData["description"] = description
	}

	// Build composite path
	cmdb.compositePath[*cmdb.workingKB] = append(cmdb.compositePath[*cmdb.workingKB], link)
	cmdb.compositePath[*cmdb.workingKB] = append(cmdb.compositePath[*cmdb.workingKB], nodeName)
	nodePath := strings.Join(cmdb.compositePath[*cmdb.workingKB], ".")

	// Check if path already exists
	if cmdb.compositePathValues[*cmdb.workingKB][nodePath] {
		return fmt.Errorf("path %s already exists in knowledge base", nodePath)
	}

	// Mark path as used
	cmdb.compositePathValues[*cmdb.workingKB][nodePath] = true

	// Store in the underlying BasicConstructDB
	path := strings.Join(cmdb.compositePath[*cmdb.workingKB], ".")
	fmt.Println("path", path)
	return cmdb.BasicConstructDB.Store(path, nodeData, nil, nil)
}

// AddInfoNode adds an info node (temporary header node that gets removed from path)
func (cmdb *ConstructMemDB) AddInfoNode(link, nodeName string, nodeData map[string]interface{}, description string) error {
	if cmdb.workingKB == nil {
		return fmt.Errorf("no working knowledge base selected")
	}

	// Add as header node first
	err := cmdb.AddHeaderNode(link, nodeName, nodeData, description)
	if err != nil {
		return err
	}

	// Remove node_name and link from path (reverse order)
	pathLen := len(cmdb.compositePath[*cmdb.workingKB])
	if pathLen >= 2 {
		cmdb.compositePath[*cmdb.workingKB] = cmdb.compositePath[*cmdb.workingKB][:pathLen-1] // Remove nodeName
		cmdb.compositePath[*cmdb.workingKB] = cmdb.compositePath[*cmdb.workingKB][:pathLen-2] // Remove link
	}

	return nil
}

// LeaveHeaderNode leaves a header node, verifying the label and name
func (cmdb *ConstructMemDB) LeaveHeaderNode(label, name string) error {
	if cmdb.workingKB == nil {
		return fmt.Errorf("no working knowledge base selected")
	}

	path := cmdb.compositePath[*cmdb.workingKB]

	// Check if path is empty
	if len(path) == 0 {
		return fmt.Errorf("cannot leave a header node: path is empty")
	}

	// Pop the name
	if len(path) < 1 {
		return fmt.Errorf("cannot leave a header node: path is empty")
	}
	refName := path[len(path)-1]
	cmdb.compositePath[*cmdb.workingKB] = path[:len(path)-1]

	// Check if we have enough elements for label
	path = cmdb.compositePath[*cmdb.workingKB]
	if len(path) == 0 {
		// Put the name back and raise an error
		cmdb.compositePath[*cmdb.workingKB] = append(cmdb.compositePath[*cmdb.workingKB], refName)
		return fmt.Errorf("cannot leave a header node: not enough elements in path")
	}

	// Pop the label
	refLabel := path[len(path)-1]
	cmdb.compositePath[*cmdb.workingKB] = path[:len(path)-1]

	// Verify the popped values
	var errorMsgs []string
	if refName != name {
		errorMsgs = append(errorMsgs, fmt.Sprintf("expected name '%s', but got '%s'", name, refName))
	}
	if refLabel != label {
		errorMsgs = append(errorMsgs, fmt.Sprintf("expected label '%s', but got '%s'", label, refLabel))
	}

	if len(errorMsgs) > 0 {
		return fmt.Errorf("assertion error: %s", strings.Join(errorMsgs, ", "))
	}

	return nil
}

// CheckInstallation checks if the installation is correct by verifying that all paths are properly reset
func (cmdb *ConstructMemDB) CheckInstallation() error {
	for kbName, path := range cmdb.compositePath {
		if len(path) != 1 {
			return fmt.Errorf("installation check failed: path is not empty for knowledge base %s. Path: %v", kbName, path)
		}
		if path[0] != kbName {
			return fmt.Errorf("installation check failed: path is not empty for knowledge base %s. Path: %v", kbName, path)
		}
	}
	return nil
}

// GetCurrentPath returns the current composite path for the working KB
func (cmdb *ConstructMemDB) GetCurrentPath() []string {
	if cmdb.workingKB == nil {
		return nil
	}
	// Return a copy to prevent external modification
	path := make([]string, len(cmdb.compositePath[*cmdb.workingKB]))
	copy(path, cmdb.compositePath[*cmdb.workingKB])
	return path
}

// GetCurrentPathString returns the current composite path as a string
func (cmdb *ConstructMemDB) GetCurrentPathString() string {
	if cmdb.workingKB == nil {
		return ""
	}
	return strings.Join(cmdb.compositePath[*cmdb.workingKB], ".")
}

// GetWorkingKB returns the currently selected working knowledge base
func (cmdb *ConstructMemDB) GetWorkingKB() *string {
	return cmdb.workingKB
}

// GetAllKBNames returns all knowledge base names
func (cmdb *ConstructMemDB) GetAllKBNames() []string {
	names := make([]string, 0, len(cmdb.compositePath))
	for name := range cmdb.compositePath {
		names = append(names, name)
	}
	return names
}

